### ISSUE TITLE
criterion uses parameters from the net

### ISSUE THREAD

**Issue by emptymalei** [skorch-dev/skorch#669]
## Some Background

I am writing a simple Mixture Density Network which has a loss function using the parameters of the net.

My `forward` method returns the parameters of the GMM,

```
def forward(self, x):

     ... some code here

    # I could also assign pi, sigma2, mu as properties of the class
    # if there is a way to get these in criterion
    # self.pi = pi
    # self. 

     return pi, sigma2, mu
```

My loss function requires these three params returned by the forward method (`pi`, `sigma2`, `mu`):

```
def mdn_loss(self, pi, sigma2, mu, target, N):
       
    ... some code here

    return loss
```

When I pass the criterion to the `NeuralNetRegressor`, I need to get these three params from the current epoch.

## Question

Is there a way to construct such a regressor with such a criterion?

## Thoughts

Looking into these lines

https://github.com/skorch-dev/skorch/blob/7a845682be15956a132c3817f0431177a620fee5/skorch/regressor.py#L36-L45

I am thinking, this might be easy if the criterion is not defaulting to `MSELoss` but take a method from the NeuralNet.

Thank you! 😄

**Comment by BenjaminBossan** [comment]
When you initialize your net, you pass your own criterion:

`net = NeuralNetRegressor(..., criterion=MyCriterion, ...)`

This ensures that your criterion is used instead of `MSELoss`.

Now, during the loss computation, the output of your module's forward call, `y_pred`, which is a tuple, is passed to the criterion:

https://github.com/skorch-dev/skorch/blob/7816c15a3b1986abd433e795b6187973f5ea9987/skorch/net.py#L1113-L1118

At that point, you should see an error, because your criterion expects 5 values to be passed, not 2. What could you do? There are two possibilities:

1. Your change your criterion's forward method to only accept 2 arguments, with the first, `y_pred`, being a tuple that contains `pi, sigma2, mu`:

```python
# inside your criterion class
def forward(self, y_pred, target):
    pi, sigma2, mu = y_pred
    # rest of the code is the same
```

2. If that's not possible or desired, you could instead override `get_loss` to unpack `y_pred`:

```python
class MyNet(NeuralNetRegressor):
    def get_loss(self, y_pred, y_true, X=None, training=False):
        y_true = to_tensor(y_true, device=self.device)

        if isinstance(self.criterion_, torch.nn.Module):
            self.criterion_.train(training)

        pi, sigma2, mu = y_pred
        return self.criterion_(pi, sigma2, mu, y_true)
```

Note that I don't know where the `N` comes from, so I ignored it in the examples.

**Comment by emptymalei** [comment]
Thank you, @BenjaminBossan 
I tweaked the second one a bit and it worked! 🎉
